src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    // This part contains all the logic for hooking up the DiagnosticAnalyzer to the CodeStyleProvider.
+    // All the code in this part is an implementation detail and is intentionally private so that
+    // subclasses cannot change anything.  All code relevant to subclasses relating to analysis
+    // is contained in AbstractCodeStyleProvider.cs
+
+    internal abstract partial class AbstractCodeStyleProvider<TOptionKind, TSyntaxKind, TCodeStyleProvider>
+    {
+        public abstract class DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+        {
+            public readonly TCodeStyleProvider _codeStyleProvider;
+
+            protected DiagnosticAnalyzer(bool configurable = true)
+                : this(new TCodeStyleProvider(), configurable)
+            {
+            }
+
+            private DiagnosticAnalyzer(TCodeStyleProvider codeStyleProvider, bool configurable)
+                : base(codeStyleProvider._descriptorId, codeStyleProvider._title, codeStyleProvider._message, configurable)
+            {
+                _codeStyleProvider = codeStyleProvider;
+            }
+
+            protected sealed override void InitializeWorker(Diagnostics.AnalysisContext context)
+                => _codeStyleProvider.DiagnosticAnalyzerInitialize(new AnalysisContext(_codeStyleProvider, context));
+
+            public sealed override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+                => _codeStyleProvider.GetDiagnosticAnalyzerCategory();
+
+            public sealed override bool OpenFileOnly(Workspace workspace)
+                => _codeStyleProvider.DiagnosticsForOpenFileOnly(workspace);
+        }
+
+        /// <summary>
+        /// Critically, we want to consolidate the logic about checking if the analyzer should run
```

Comment:
Ok.  So where does that get done?  And how is it done.  If someone doesn't know about options, how do they know what sort of diagnostic to produce (since it is option-value dependent)?

Reply:
The implementation here is a unification of behavior we have today. I'm interested in alternatives such as #27712 but I see no reason why that open design work would block forward progress of this as a clarifying/supporting change.